staker/txbuilder/builder.go

```diff
@@ -30,34 +28,76 @@ type ValidatorWalletInterface interface {
 // This inherits from an ethclient.Client so it can be used to transparently
 // intercept calls to SendTransaction and queue them for the next batch.
 type Builder struct {
-	*ethclient.Client
 	transactions []*types.Transaction
-	builderAuth  *bind.TransactOpts
+	singleTxAuth bind.TransactOpts
+	multiTxAuth  bind.TransactOpts
 	isAuthFake   bool
+	authMutex    sync.Mutex
 	wallet       ValidatorWalletInterface
+	gasRefunder  common.Address
 }
 
-func NewBuilder(wallet ValidatorWalletInterface) (*Builder, error) {
-	randKey, err := crypto.GenerateKey()
-	if err != nil {
-		return nil, err
-	}
-	builderAuth := wallet.AuthIfEoa()
+func NewBuilder(wallet ValidatorWalletInterface, gasRefunder common.Address) (*Builder, error) {
+	var builderAuth bind.TransactOpts
 	var isAuthFake bool
-	if builderAuth == nil {
-		// Make a fake auth so we have txs to give to the smart contract wallet
-		builderAuth, err = bind.NewKeyedTransactorWithChainID(randKey, big.NewInt(9999999))
+	if authIfEoa := wallet.AuthIfEoa(); authIfEoa != nil {
+		builderAuth = *authIfEoa
+	} else {
+		isAuthFake = true
+		var addressOrZero common.Address
+		if addr := wallet.Address(); addr != nil {
+			addressOrZero = *addr
+		}
+		builderAuth = bind.TransactOpts{
+			From:     addressOrZero,
+			GasLimit: 123, // don't gas estimate, that's done when the real tx is created
+			Signer: func(_ common.Address, tx *types.Transaction) (*types.Transaction, error) {
+				return tx, nil
+			},
+		}
+	}
+	builderAuth.NoSend = true
+	builder := &Builder{
+		singleTxAuth: builderAuth,
+		multiTxAuth:  builderAuth,
+		wallet:       wallet,
+		isAuthFake:   isAuthFake,
+		gasRefunder:  gasRefunder,
+	}
+	originalSigner := builderAuth.Signer
+	builder.multiTxAuth.Signer = func(addr common.Address, tx *types.Transaction) (*types.Transaction, error) {
+		tx, err := originalSigner(addr, tx)
 		if err != nil {
 			return nil, err
 		}
-		isAuthFake = true
+		// Append the transaction to the builder's queue of transactions
+		builder.transactions = append(builder.transactions, tx)
+		err = builder.wallet.TestTransactions(context.TODO(), builder.transactions)
```

Comment:
I just want to register that I'm not loving the context.TODO sprinkled around the code. I noticed that the auth in the NewDataPoster is also just injecting a context.TODO in it's singer function.  Of course, it feels like the "original sin" here is that bind.TransactOpts captures the network context and stores it as a field. This is pretty strongly discouraged. But, I guess we have to make the most of it.

Reply:
Yeah agreed. Elsewhere we use a custom signer function that takes in a context, but go-ethereum abi bind doesn't support that. Unfortunately the only other option is having the tx builder intercept RPC requests, which seems worse to me.

Reply:
Again, I wonder if context.Background() is a better choice here. It's a shame that the Signer function doesn't have a reference to its parent TransactOpts. If it did, we could use the txOpts.Context field which "should" be populated by the time it's being signed, right? I think we can't just use builder.multiTxAuth.Context because that isn't actually mutated (by design) in calls to `Auth` or `AuthWithAmount`.

Hmmm. Maybe context.Background? The reason I'm suggesting that is, I don't see a path for ever resolving this TODO. It feels like we'll never have a Request from which to grab a Context, and we don't really expect to change how go-ethereum works. So, I think our only remaining option is to use the context of the process.

Reply:
The two main ways I could see to resolve this for bold are:
- The signer returns the original tx unsigned, and then the bold transactor transforms that with its context. The advantage of this is that it gets the context, but the disadvantage is that we have a sort of meaningless tx floating around. E.g. if in bold we tried to wait for that transaction to get confirmed, we'd get stuck because that tx never actually gets posted on-chain, it just gets transformed into the smart contract wallet tx.
- Geth is working on abigen v2 which I believe would split out a lot of the stuff we currently have to put into Signer and hopefully resolve this. Speaking of which, we should look into the status of that. It's definitely a longer term thing but I don't think we'll have this TODO forever, hopefully

Reply:
Okay. We can leave it as TODO for now. Resolving.